### ISSUE TITLE
Table fields not remembering their Default Values row order

### ISSUE THREAD

**Issue by olivierbon** [craftcms/cms#3947]
### Description

Table fields don’t remember changes to the row order in their Default Values setting.

### Steps to reproduce

1. Create a Table field
2. Add multiple rows to the Default Values setting
3. Change the order of the Default Values rows
4. Save the field

The rows will be saved in the order they were created in, rather than the custom-defined row order.

### Additional info

- Craft version: 3.1.15
- PHP version: 7.2.1

**Comment by narration-sd** [comment]
Brandon, I'm still seeing this problem, on dev-develop -- just checked again this evening.

Nothing's changed in my source definition, and I don't see any settings to alter ordering from the physical order rows are placed on (in Craft CP - My Plugin Settings)

If I move a row in the usual way (ordering is critical for this plugin), and save, it flips back to its original position. 

Old rows do stay in their original positions, so I noticed this first on an added row. But if I move old rows, they also are 'stuck' to their pre-move positions.

Should be simple, I think -- needed for a beta next week...thanks.

**Comment by brandonkelly** [comment]
@narration-sd this bug was specifically for the Default Values setting on Table fields. Is that where you’re seeing it too? I verified that the fix worked. Maybe Composer‘s cache was out of date. Rather than being on `dev-develop` you should try setting the `craftcms/cms` requirement to `3.1.16` or `^3.1.16`, then run `composer update`.

**Comment by narration-sd** [comment]
Fair enough, and thanks for the quick attention. I'll run through your scenarios and come back...

**Comment by narration-sd** [comment]
Ok, I've ditched /vendor, cleared composer caches, set Craft to ^3.1.16 (appropriate actually towards the delivery this is for), cleaned out non-essential plugins, and composer install'd.

Moving table rows in my plugin settings still reverts on save. Value changes seem to save fine.

Brandon, is there anything I could do that you would imagine to break ordering?

I did add some items to this plugin/settings recently, if none of them were in the table. I'm going to try reverting to some earlier version to see if re-ordering starts functioning again...

What happens ever when you promise something soonish, and I still have to write that code ;)

**Comment by narration-sd** [comment]
Ok!  It's Craft latest or so causing this.

- I ran my own plugin back to a stone-age version where it's sure ordering has worked afterwards
-- result, no change
- I ran back Craft as far as a locked Yii update allowed, 3.1.3
-- result: success!  Manual ordering and saving settings from the table functions as expected

Maybe the fix you put in on this original issue made the breakage?

Happy to enter this as a separate issue if not/or if you prefer...and I will go back to using craftcms/cms:dev-master to validate your fix towards release...

**Comment by brandonkelly** [comment]
> Moving table rows in my plugin settings still reverts on save. Value changes seem to save fine.

Once again this issue was specifically for the Default Values setting of table fields. if you’re seeing it _anywhere else_ then it’s unrelated, though the fix will likely be similar. It happens because associative arrays are sorted by their keys when saved to the project config, in an effort to reduce merge conflicts / unexpected changes in project.yaml.

So when saving an array where the order matters, it should be numerically indexed, rather than associative. You can do that by adding this to your component/model’s `init()` method:

```php
if (!empty($this->settingName)) {
    $this->settingName = array_values($this->settingName);
}
```

Where `settingName` is the property name holding the editable table row data.

**Comment by narration-sd** [comment]
Ok, I see what you're getting at, but this isn't working, as the Settings model isn't filled in at init() time, even though I've called parent::init() prior to invoking this code.

Should I be doing something in the plugin's own init() that I'm overlooking?

(smiling because I got bitten by project config, even while waiting on it here)

**Comment by narration-sd** [comment]
also reverted to 3.1.17 after noticing that, in case influenced though didn't think so.  It doesn't.

feeling is I need to call something within SettingsModel::init() besides parent::init() to get the values filled in. Rather than probe, guess this time, asking. 

As this is playing hob with needed gross refactoring of a complex React part ;)

Thanks again. Saved by mexican music, with wit. Like one? 
@brandonkelly

**Comment by brandonkelly** [comment]
For plugin settings you could do it from `setAttributes()`.

```php
public function setAttributes($values, $safeOnly = true)
{
    parent::setAttributes($values, $safeOnly);
    if (!empty($this->settingName)) {
        $this->settingName = array_values($this->settingName);
    }
}
```

**Comment by narration-sd** [comment]
...does the trick.  Thanks very kindly, Brandon - the day is saved. Rest of it anyway, but you know what I mean with an exposure coming up.

Here's the lady, considerably intelligent, creative, thoughtful, in case you like it.  For the wit, besides the song delivery, just watch the video, right up to the last moment :)

- https://www.youtube.com/watch?v=u0mHmduy_Ek
- https://www.youtube.com/watch?v=E6wuWc6w9D8

Well, two. The second one is kind of a Roma treatment...if the listening is the thing.

You might guess I like her. Not least for the great variety of music across the stages of her building. From the German, a true bildungsroman.

Cheers, Brandon, and on your good deed for the day.
Clive
